feat(dot/sync): implement codeSubstitutes

edwardmack opened the pull request:
## Changes

<!--

Please provide a brief but specific list of changes made, describe the changes
in functionality rather than the changes in code.

-->

- Implement code to handle code substitutes defined in genesis.json. wasm code is replaced with code contained in codeSubstitutes map when block is imported with same hash as key in codeSubstitutes map.
- (Note, this PR replaces PR#1620 which was closed due to issues with CLA check).
-

## Tests

<!--

Details on how to run tests relevant to the changes within this pull request.

-->

```
go test ./dot/sync/... -v -run TestSyncer_H
```

## Issues

<!--

Please link any issues that this pull request is related to and use the GitHub
supported format for automatically closing issues (ie, closes #123, fixes #123)

-->

- closes #1619

EclesioMeloJunior reviewed (approved):
lgtm :D

noot commented on dot/sync/syncer.go:
```diff
@@ -424,6 +439,27 @@ func (s *Service) handleRuntimeChanges(newState *rtstorage.TrieState) error {
 
+	codeSubBlockHash, err := s.codeSubstitutedState.LoadCodeSubstitutedBlockHash()
+	if err != nil {
+		logger.Debug("problem checking code substituted block hash", "error", err)
```
return err

edwardmack replied:
I updated codeSubstitutedState.LoadCodeSubstitutedBlockHash does not return error.

noot commented on dot/sync/syncer.go:
```diff
@@ -424,6 +439,27 @@ func (s *Service) handleRuntimeChanges(newState *rtstorage.TrieState) error {
+		//  (do a runtime change if code substituted and runtime spec versions are different, or code not substituted)
+		newVersion, err := s.runtime.CheckRuntimeVersion(code) // nolint
+		if err != nil {
+			logger.Debug("problem checking runtime version", "error", err)
```
return err

edwardmack replied:
updated.

noot commented on dot/sync/syncer.go:
```diff
@@ -434,6 +470,31 @@ func (s *Service) handleRuntimeChanges(newState *rtstorage.TrieState) error {
 	return nil
 }
 
+func (s *Service) handleCodeSubstitution(block common.Hash) error {
```
rename to `hash`

edwardmack replied:
renamed.

noot commented on dot/sync/syncer.go:
```diff
@@ -424,6 +439,27 @@ func (s *Service) handleRuntimeChanges(newState *rtstorage.TrieState) error {
+			"previous code hash", s.codeHash, "new code hash", currCodeHash,
+			"previous spec version", previousVersion.SpecVersion(), "new spec version", newVersion.SpecVersion())
+	}
+
```
you will also need to do `StoreCodeSubstitutedBlockHash(common.Hash{})` here to store in the db that no code substitution is currently happening, otherwise on restart the node will still think there's a code substitution

edwardmack replied:
Good catch, updated.

noot commented on lib/runtime/wasmer/instance.go:
```diff
@@ -216,6 +216,50 @@ func (in *Instance) UpdateRuntimeCode(code []byte) error {
+
+	instance.SetContextData(in.ctx)
+
+	in.vm = instance
+	checkVersion, err := in.Version()
```
not sure why this is needed, you can just call `instance.Version()` directly and remove the `originalVM := in.vm` and `in.vm = originalVM` lines

edwardmack replied:
I can't call `instance.Version()`  because instance is a wasmer instance which doesn't have Version() function.

noot commented on lib/runtime/wasmer/instance.go:
```diff
@@ -216,6 +216,50 @@ func (in *Instance) UpdateRuntimeCode(code []byte) error {
 }
 
+// CheckRuntimeVersion calculates runtime Version for runtime blob passed in
+func (in *Instance) CheckRuntimeVersion(code []byte) (runtime.Version, error) {
```
this function has code that's mostly the same as `UpdateRuntimeVersion`, could you put the code that overlaps into a separate function and call that in both?

edwardmack replied:
created `setupInstanceVM` function to handle redundant code.

noot reviewed (commented):
tested with the network and it works :D functionality is working, just some code comments. also would be good not to change the gssmr genesis file, is there a way to use some test genesis instead or just modify the `genesisData` that's used to have a `codeSubstitution` field?

also please add this above `syncer.go` line 229 (inside the `if hasHeader && hasBody` case):
```
if err := s.handleCodeSubstitution(bd.Hash); err != nil {
    logger.Warn("failed to handle code substitution", "error", err)
    return i, err
}
```

this makes sure that the code substitution happens even when blocks are already stored in the db, in case the node crashed or used `--rewind`. had to add this to test out the changes since i had to rewind the chain

edwardmack commented on dot/services.go:
```diff
@@ -95,6 +96,23 @@ func createRuntime(cfg *Config, st *state.Service, ks *keystore.GlobalKeystore,
+		logger.Debug("failed to retrieve substituted code block hash", "error", err)
+	}
+
+	if !reflect.DeepEqual(codeSubHash, common.Hash{}) {
```
Good call, I didn't realize common.Hash had an Equal function.  Updated.

edwardmack commented on dot/services.go:
```diff
@@ -377,16 +395,26 @@ func createBlockVerifier(st *state.Service) (*babe.VerificationManager, error) {
 }
 
 func createSyncService(cfg *Config, st *state.Service, bp sync.BlockProducer, fg sync.FinalityGadget, dh *core.DigestHandler, verifier *babe.VerificationManager, rt runtime.Instance) (*sync.Service, error) {
```
updated.

edwardmack commented:
I think I've addressed all comments.  I've removed codeSubstitutes from gossamer.json genesis, and modified test helper NewTestSyncer to take a boolean usePolkadotGenesis flag, which when true uses polkadot genesis, when false uses gossamer genesis.

noot reviewed (approved):
great work!
